[jit][edge] Export maybe-used interface methods from modules.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65972
* #65971
* #65970
* #65969
* #65968
* #65967
* __->__ #65966





Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

zhxchen17 committed:
[jit][edge] Export maybe-used interface methods from modules.



Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."


Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

qihqi reviewed (approved):
overall looks pretty solid, thanks!

iseeyuan reviewed (commented):
Thanks! As discussed offline, 
1. Please make sure that writing the new features or parts (like interfaces) are disabled by default. We should enable the writing part only after the reading part is enabled and the flow is tested.
2. Please add tests for debug stack trace when inline is not used, to make sure we still have the save exception message before and after.

Also a couple of suggestions on a relatively "big" PR like this.
1. Add some description in the summary, like the purpose of this PR (introduced exportSet to have both top/nontop, to collect the interface calls and add them to the exportSet, etc.), testing plan, todo list in the followup PRs, etc.
2. There are also some pure refactoring changes (like moving codes from one file to another. Consider split the PR to two. One is the code movement only, and the second is the actual update. It would save a lot of review time and accelerate the iterations.

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit][edge] Export maybe-used interface methods from modules."






Support exportation of "interface methods" from submodule to a mobile module. "Interface methods" are defined as methods which might be dynamically called in a module therefore need to be exported anyway, like virtual functions in C++.

Before this change the algorithm of exportation is a simple iteration through all toplevel methods. Now since we have indirect calls, we need to recursively walkthrough the call graph to find all potentially used methods, which means the order we export methods might break in old runtimes, to guarantee forward compatibility we need to export toplevel methods first, then extra methods, in this order toplevel methods will always be found first.

NOTE that interface methods exportations are disabled by default in this diff. We need to call torch._C._enable_mobile_interface_call_export to actaully enable it.

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31326155/)!

Differential Revision: [D31326155](https://our.internmc.facebook.com/intern/diff/D31326155)

[ghstack-poisoned]
